### ISSUE TITLE
Granting access for any app [desktop client, android app] results in an internal server error when served under a different webroot

### ISSUE THREAD

**Issue by dosera** [nextcloud/server#28965]
### Steps to reproduce
1. Setup nextcloud in docker behind a reverse proxy and serve it under a basepath e.g. `mydomain.com/nextcloud`
2. Try to grant access to some application (e.g. android app, nextcloud desktop client)

### Expected behaviour
Access should be granted.

#### Issue

Granting access for app (nextcloud-client on linux, android app) with 

`fastcgi_param front_controller_active true;`

enabled for pretty urls in a 

`nginx -> docker setup`

served under a different webroot (`mydomain:1234/nextcloud`) results in a 500 and an exception on the server (see below)
    
Everything else works fine.

#### Possibly related issue
Unsure, but I found a solution here posted in the comment by okapia: https://github.com/nextcloud/server/issues/23117

### Actual behaviour
Internal server error.

### Server configuration

**Operating system:** Arch Linux

**Web server:** nginx

**Database:** mariadb 10

**PHP version:** PHP 7.4.24 (cli) (built: Sep 23 2021 21:24:59) ( NTS )

**Nextcloud version:** (see Nextcloud admin page) 22.1.1

**Updated from an older Nextcloud/ownCloud or fresh install:** -

**Where did you install Nextcloud from:** linuxserver.io docker image

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
    'overwrite.cli.url' => 'https://mydomain:1234/nextcloud',
    'overwritehost' => 'mydomain:1234',
    'overwriteprotocol' => 'https',
    'overwritewebroot' => '/nextcloud',
    'trusted_proxies' =>   
      array (     
        0 => '192.168.1.1',
      ),
```
</details>

**Reverse Proxy configuration:**
<details>
<summary>Config report</summary>

    The 'outer' nginx is configured to forward accordingly to the docker container, e.g.

        location /nextcloud {
            proxy_pass https://localhost:10000;
            proxy_set_header Connection 'upgrade';
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header X-Content-Type-Options nosniff;
            proxy_set_header X-XSS-Protection "1; mode=block";
            proxy_set_header X-Robots-Tag none;
            proxy_set_header X-Download-Options noopen;
            proxy_set_header X-Frame-Options "SAMEORIGIN";
            proxy_set_header X-Permitted-Cross-Domain-Policies none;
            proxy_set_header Referrer-Policy no-referrer;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Path '/nextcloud';
            client_max_body_size 2G;
        }

    The 'inner' nginx is configured as is described in the docs.
</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend:** no

### Client configuration
* Nextcloud Desktop Client (linux, version 3.3.3-2)
* Nextcloud Android (most recent)

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
Request to:

        https://mydomain:1234/nextcloud/login/v2/grant?stateToken=tjINkG8bI4GqOHjB6tLBvgyK8lLa5jIC2qn81Q4J25PuLv3L0i6QNzc7UXFC6fkT

Log:

    2021/09/21 20:19:13 [error] 413#413: *67 FastCGI sent in stderr: "PHP message: [owncloud][files_sharing][0] /appinfo/app.php is deprecated, use \OCP\AppFramework\Bootstrap\IBootstrap on the application class instead.PHP message: [owncloud][index][3]
    {
      "Code": 0,
      "Exception": "Exception",
      "File": "/config/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
      "Line": 158,
      "Message": "Argument 1 passed to OC\\Core\\Controller\\ClientFlowLoginV2Controller::grantPage() must be of the type string, null given, called in /config/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php on line 217",
      "Previous": {
        "Exception": "TypeError",
        "Message": "Argument 1 passed to OC\\Core\\Controller\\ClientFlowLoginV2Co\" while reading response header from upstream, client: 192.168.1.1, server: _, request: \"GET /nextcloud/login/v2/grant?stateToken=tjINkG8bI4GqOHjB6tLBvgyK8lLa5jIC2qn81Q4J25PuLv3L0i6QNzc7UXFC6fkT HTTP/1.0\", upstream: \"fastcgi://127.0.0.1:9000\", host: \"localhost:10000\""
      },
      "Trace": [
        {
          "class": "OC\\AppFramework\\Http\\Dispatcher",
          "file": "/config/www/nextcloud/lib/private/AppFramework/App.php",
          "function": "dispatch",
          "line": 156,
          "type": "->"
        },
        {
          "class": "OC\\AppFramework\\App",
          "file": "/config/www/nextcloud/lib/private/Route/Router.php",
          "function": "main",
          "line": 301,
          "type": "::"
        },
        {
          "class": "OC\\Route\\Router",
          "file": "/config/www/nextcloud/lib/base.php",
          "function": "match",
          "line": 1000,
          "type": "->"
        },
        {
          "class": "OC",
          "file": "/config/www/nextcloud/index.php",
          "function": "handleRequest",
          "line": 36,
          "type": "::"
        }
      ]
    }


```
</details>

**Comment by LukeLR** [comment]
I had a similar issue right now. Nextcloud v24.0.0 is hosted in a container with nginx v1.18.0 as webserver, behind another nginx v1.18.0 as reverse proxy. I added

'overwritewebroot' => '/',
'overwriteprotocol' => 'https',
'overwritehost' => '<external hostname>',

as instructed by the Nextcloud administration overview. However, afterwards authentication wouldn't work, with an error message similar to the above. Removing the above parameters would just result in timeouts. However, keeping them as is and adding 'forcessl' => true fixed the issue. front_controller_active does not seem to be enabled in my case. At least in my case, #23117 does indeed seem to be related.

**Comment by dosera** [comment]
Well the issue here (and indicated by the comment in #23117) is that this is a problem **with** fastcgi_param front_controller_active true; enabled.
Without this it works as is written over there and here.

**Comment by szaimen** [comment]
Hi, please update to 24.0.9 or better 25.0.3 and report back if it fixes the issue. Thank you!

My goal is to add a label like e.g. 25-feedback to this ticket of an up-to-date major Nextcloud version where the bug could be reproduced. However this is not going to work without your help. So thanks for all your effort!

If you don't manage to reproduce the issue in time and the issue gets closed but you can reproduce the issue afterwards, feel free to create a new bug report with up-to-date information by following this link: https://github.com/nextcloud/server/issues/new?assignees=&labels=bug%2C0.+Needs+triage&template=BUG_REPORT.yml&title=%5BBug%5D%3A+

**Comment by dosera** [comment]
The issue still occurs:

```
2023/01/23 22:25:09 [error] 200#200: *189 FastCGI sent in stderr: "PHP message: [owncloud][index][3] 
{
    "Code": 0,
    "Exception": "Exception",
    "File": "/config/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
    "Line": 165,
    "Message": "OC\\Core\\Controller\\ClientFlowLoginV2Controller::grantPage(): Argument #1 ($stateToken) must be of type string, null given, called in /config/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php on line 225 in file '/config/www/nextcloud/core/Controller/ClientFlowLoginV2Controller.php' line 152",
    "Previous":
    {
        "Exception": "TypeError",
        "Message": "Exception while reading response header from upstream, client: 192.168.1.1, server: _, request: \"GET /nextcloud/login/v2/grant?user=&stateToken=GuQRZvRXjff5xVMnrbuOq79pAPX7sZS03JE69l0duSNtvsnCBmIYb1wJYInPx1D5 HTTP/1.0\", upstream: \"fastcgi://127.0.0.1:9000\", host: \"127.0.0.1:10000\""
    },
    "Trace":
    [
        {
            "class": "OC\\AppFramework\\Http\\Dispatcher",
            "file": "/config/www/nextcloud/lib/private/AppFramework/App.php",
            "function": "dispatch",
            "line": 172,
            "type": "->"
        },
        {
            "class": "OC\\AppFramework\\App",
            "file": "/config/www/nextcloud/lib/private/Route/Router.php",
            "function": "main",
            "line": 298,
            "type": "::"
        },
        {
            "class": "OC\\Route\\Router",
            "file": "/config/www/nextcloud/lib/base.php",
            "function": "match",
            "line": 1047,
            "type": "->"
        },
        {
            "class": "OC",
            "file": "/config/www/nextcloud/index.php",
            "function": "handleRequest",
            "line": 36,
            "type": "::"
        }
    ]
}
```

also the logmessage seems not to be valid json anymore - there is something wrong around the "Previous" part (manually modified in the json above):
```
..
"Previous":{"Excep" while reading response ...
..
```

**Comment by szaimen** [comment]
Which nc version did you test?

**Comment by dosera** [comment]
> Which nc version did you test?

just edited it in: 25.0.3.2

**Comment by joshtrichards** [comment]
Have you seen this happen lately? Last known bug I see in this area was addressed in 25.0.5. e.g. see #32387

**Comment by nextcloud-command** [comment]
This issue has been automatically marked as stale because it has not had recent activity and seems to be missing some essential information. It will be closed if no further activity occurs. Thank you for your contributions.
